### PR TITLE
STRY50557163 : [Delta-Loading]Find the delta of changes between currently loaded application files and new changes applied while Import from SC

### PR DESCRIPTION
**Pull request by akkinenic**
[Delta-Loading]Find the delta of changes between currently loaded application files and new changes applied while Import from SC

This is a customization in the JGIT to filter out the noise and reduce the source control app install on the service now instance.

The customization ONLY includes changes to DiffCommand and DiffFormatter and we exposed a new method and the purpose is to not break existing JGIT in any way.

### CODE REVIEW

File: org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java [hunk 1]
```diff
@@ -651,6 +657,48 @@ public void format(List<? extends DiffEntry> entries) throws IOException {
...
+	 * The filter acts only for files that have MODIFY change Type.
+	 * If there are no changes detected, we will remove the diff entry.
+	 * @param entries
+	 * @param deltaFilterPattern
+	 * @throws IOException
+	 */
+
+	public void format(List<? extends DiffEntry> entries, Pattern deltaFilterPattern) throws IOException {
+		if (deltaFilterPattern == null)
+			format(entries);
```
Comment (bmeeder22) [inline org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java:671]: Also empty string ("")? also make a void return after this and we can eliminate the else block.

File: org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java [hunk 2]
```diff
@@ -651,6 +657,48 @@ public void format(List<? extends DiffEntry> entries) throws IOException {
...
+		else {
+
+			Iterator<? extends DiffEntry> diIterator = entries.iterator();
+
+			while (diIterator.hasNext()) {
+
+				DiffEntry diffEntry = diIterator.next();
+
+				if (MODIFY.equals(diffEntry.changeType)) {
```
Comment (bmeeder22) [inline org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java]: Don't need newlines between ALL of these lines... Makes the code harder to read.

File: org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java [hunk 3]
```diff
@@ -651,6 +657,48 @@ public void format(List<? extends DiffEntry> entries) throws IOException {
...
+					String bContent = new String(res.b.content);
+
+					aContent = deltaFilterPattern.matcher(aContent).replaceAll(EMPTY_STRING);
+					bContent = deltaFilterPattern.matcher(bContent).replaceAll(EMPTY_STRING);
+
+					if (!aContent.equals(bContent)) {
+						format(res.header, res.a, res.b);
+					} else {
+						diIterator.remove();
+					}
```
Comment (bmeeder22) [inline org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java]: Break this into a private helper method.

**Review by bmeeder22 — CHANGES_REQUESTED** [review]

File: org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java (hunk 3 above)
Comment (akkinenic) [reply org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java]: ah, I can do it. But its just 6 lines and it's already calling createFormatResult and  format methods.

File: org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java (hunk 2 above)
Comment (akkinenic) [reply org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java]: :)

File: org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java (hunk 1 above)
Comment (akkinenic) [reply org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java:671]: deltaFilterPattern is an object. And if the pattern is null, it goes to the usual Jgit flow which is the format(entries).

Comment (bmeeder22) [reply org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java:671]: Ok, then just do the `return;` after so we can eliminate the else block.

File: org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java (hunk 3 above)
Comment (bmeeder22) [reply org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java]: I get it, just makes it easier to read, makes your higher level function more clear in its actual intent.

File: org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java (hunk 1 above)
Comment (akkinenic) [reply org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java:671]: changed it

File: org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java (hunk 2 above)
Comment (akkinenic) [reply org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java]: changed it

File: org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java (hunk 3 above)
Comment (akkinenic) [reply org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java]: leaving in the same method

File: org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java [hunk 4]
```diff
@@ -651,6 +657,41 @@ public void format(List<? extends DiffEntry> entries) throws IOException {
...
+				String aContent = new String(res.a.content);
+				String bContent = new String(res.b.content);
+				aContent = deltaFilterPattern.matcher(aContent).replaceAll(EMPTY_STRING);
+				bContent = deltaFilterPattern.matcher(bContent).replaceAll(EMPTY_STRING);
+				if (!aContent.equals(bContent)) {
+					format(res.header, res.a, res.b);
+				} else {
+					diIterator.remove();
+				}
+				} else {
```
Comment (bmeeder22) [inline org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java]: Formatting issue on this line.

**Review by bmeeder22 — APPROVED** [review]
Fix the formatting otherwise good.

File: org.eclipse.jgit/src/org/eclipse/jgit/api/DiffCommand.java [hunk 5]
```diff
@@ -94,13 +96,17 @@
 
 	private ProgressMonitor monitor = NullProgressMonitor.INSTANCE;
 
+	private Pattern deltaFilterPattern = null;
```
Comment (axfiori) [inline org.eclipse.jgit/src/org/eclipse/jgit/api/DiffCommand.java:99]: can you add javadoc and mention that the filter is not going to be used if showNameAndStatusOnly is set to true

File: org.eclipse.jgit.test/tst/org/eclipse/jgit/api/DiffCommandTest.java [hunk 6]
```diff
@@ -241,6 +242,31 @@ public void testNoOutputStreamSet() throws Exception {
 		assertEquals("test.txt", diff.getNewPath());
 	}
 
+	@Test
+	/**
+	 * Setting a delta filter of type ANY (applies to any file) with regex that matches the
```
Comment (axfiori) [inline org.eclipse.jgit.test/tst/org/eclipse/jgit/api/DiffCommandTest.java]: can you update the description? it is from the time in which we had filter types. ty

File: org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java (hunk 4 above)
Comment (akkinenic) [reply org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java]: fixed it

File: org.eclipse.jgit/src/org/eclipse/jgit/api/DiffCommand.java (hunk 5 above)
Comment (akkinenic) [reply org.eclipse.jgit/src/org/eclipse/jgit/api/DiffCommand.java:99]: added javadoc

File: org.eclipse.jgit.test/tst/org/eclipse/jgit/api/DiffCommandTest.java (hunk 6 above)
Comment (akkinenic) [reply org.eclipse.jgit.test/tst/org/eclipse/jgit/api/DiffCommandTest.java]: updated.

**Review by mkwspam33 — APPROVED** [review]

File: org.eclipse.jgit.test/tst/org/eclipse/jgit/api/DiffCommandTest.java [hunk 7]
```diff
@@ -241,6 +242,31 @@ public void testNoOutputStreamSet() throws Exception {
...
+		write(new File(folder, "folder.txt"), "folderchange");
+
+		OutputStream out = new ByteArrayOutputStream();
+		Pattern deltaFilterPattern = Pattern.compile("change");
+		List<DiffEntry> entries = git.diff().setDeltaFilterPattern(deltaFilterPattern).setOutputStream(out).call();
+		assertEquals(0, entries.size());
+
+		String actual = out.toString();
+		String expected = "";
+		assertEquals(expected.toString(), actual);
```
Comment (wendts) [inline org.eclipse.jgit.test/tst/org/eclipse/jgit/api/DiffCommandTest.java]: Make these 3 a single line - assertEquals("", out.toString());

File: org.eclipse.jgit.test/tst/org/eclipse/jgit/api/DiffCommandTest.java [hunk 8]
```diff
@@ -241,6 +242,31 @@ public void testNoOutputStreamSet() throws Exception {
...
+	}
```
Comment (wendts) [inline org.eclipse.jgit.test/tst/org/eclipse/jgit/api/DiffCommandTest.java:266]: Should we replace all of the strings in this method (and following methods) with constants?

File: org.eclipse.jgit.test/tst/org/eclipse/jgit/diff/DiffFormatterTest.java [hunk 9]
```diff
@@ -343,6 +344,190 @@ public void testDiff() throws Exception {
...
+
+		//testing a delta filter with one regex (ANY)
+		Pattern deltaFilterPattern = Pattern.compile("change");
+		dfmt.format(dfmt.scan(oldTree, newTree), deltaFilterPattern);
+		dfmt.flush();
+
+		String actual = os.toString("UTF-8");
+		String expected = "";
+
+		assertEquals(expected, actual);
```
Comment (wendts) [inline org.eclipse.jgit.test/tst/org/eclipse/jgit/diff/DiffFormatterTest.java]: Combine these - assertEquals("", os.toString("UTF-8"));

File: org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java [hunk 10]
```diff
@@ -651,6 +657,41 @@ public void format(List<? extends DiffEntry> entries) throws IOException {
...
+				FormatResult res = createFormatResult(diffEntry);
+				String aContent = new String(res.a.content);
+				String bContent = new String(res.b.content);
+				aContent = deltaFilterPattern.matcher(aContent).replaceAll(EMPTY_STRING);
+				bContent = deltaFilterPattern.matcher(bContent).replaceAll(EMPTY_STRING);
+				if (!aContent.equals(bContent)) {
+					format(res.header, res.a, res.b);
+				} else {
+					diIterator.remove();
+				}
```
Comment (wendts) [inline org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java]: No need for the curly braces around single lines

Comment (akkinenic) [reply org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java]: oh yeah, I keep forgetting to not add those braces. style guide :(.

File: org.eclipse.jgit.test/tst/org/eclipse/jgit/api/DiffCommandTest.java (hunk 8 above)
Comment (akkinenic) [reply org.eclipse.jgit.test/tst/org/eclipse/jgit/api/DiffCommandTest.java:266]: Most of the format is the original Jgit tests format, intended to keep that way.

File: org.eclipse.jgit.test/tst/org/eclipse/jgit/api/DiffCommandTest.java (hunk 7 above)
Comment (akkinenic) [reply org.eclipse.jgit.test/tst/org/eclipse/jgit/api/DiffCommandTest.java]: fixed it

File: org.eclipse.jgit.test/tst/org/eclipse/jgit/diff/DiffFormatterTest.java (hunk 9 above)
Comment (akkinenic) [reply org.eclipse.jgit.test/tst/org/eclipse/jgit/diff/DiffFormatterTest.java]: changed it

File: org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java (hunk 10 above)
Comment (akkinenic) [reply org.eclipse.jgit/src/org/eclipse/jgit/diff/DiffFormatter.java]: changed it

**Review by wendts — APPROVED** [review]

**Pull request merged**